whisper/tokenizer.py

```diff
@@ -127,21 +129,47 @@
 }
 
 
-@dataclass(frozen=True)
+@dataclass
 class Tokenizer:
     """A thin wrapper around `GPT2TokenizerFast` providing quick access to special tokens"""
 
-    tokenizer: "GPT2TokenizerFast"
-    language: Optional[str]
-    sot_sequence: Tuple[int]
+    encoding: tiktoken.Encoding
+    language: Optional[str] = None
+    task: Optional[str] = None
+    sot_sequence: Tuple[int] = ()
+    special_tokens: Dict[str, int] = field(default_factory=dict)
+
+    def __post_init__(self):
+        for special in self.encoding.special_tokens_set:
+            special_token = self.encoding.encode_single_token(special)
+            self.special_tokens[special] = special_token
+
+        sot: int = self.special_tokens["<|startoftranscript|>"]
+        translate: int = self.special_tokens["<|translate|>"]
+        transcribe: int = self.special_tokens["<|transcribe|>"]
+
+        langs = tuple(LANGUAGES.keys())
+        sot_sequence = [sot]
+        if self.language is not None:
+            sot_sequence.append(sot + 1 + langs.index(self.language))
+        if self.task is not None:
+            task_token: int = transcribe if self.task == "transcribe" else translate
+            sot_sequence.append(task_token)
+
+        self.sot_sequence = tuple(sot_sequence)
 
     def encode(self, text, **kwargs):
-        return self.tokenizer.encode(text, **kwargs)
+        return self.encoding.encode(text, **kwargs)
 
     def decode(
         self, token_ids: Union[int, List[int], np.ndarray, torch.Tensor], **kwargs
     ):
-        return self.tokenizer.decode(token_ids, **kwargs)
+        if isinstance(token_ids, int):
+            token_ids = [token_ids]
+        if isinstance(token_ids, (torch.Tensor, np.ndarray)):
+            token_ids = token_ids.tolist()
```

Comment:
```suggestion
```

Reply:
All calls to decode() in the code pass a List[int] argument. Moreover, callers can always care to pass a list. 
These "if"-tests cost time which spoil the superior runtime of tiktoken.
If the tests should remain then use 
```
if isinstance(token_ids, list):
    pass
elif isinstance(token_ids, int):
    token_ids = [token_ids]
elif isinstance(token_ids, (torch.Tensor, np.ndarray)):
    token_ids = token_ids.tolist()
```
 This makes just one test if the argument is List[int]

Reply:
regarding [ `decode_with_timestamps()`](https://github.com/openai/whisper/pull/1044/files#diff-c4ef7e42e8edd6ba1a400cc0fdcb54fccd47a4629979fe66108285a5c3a58053R174) in `tokenizer.py`, why not add the timestampt strings to the tokenizer as special tokens? sth like 
"<|0.00|>", self.timestamp_begin
....
"<|30.00|>", self.timestamp_begin+1500

Reply:
Great points; reflected in the commits below.